File: scripts/migration/add-root-bounds.swift
import Foundation

let repo = "/Users/vishnudutt/layout-test"
let resources = "\(repo)/Sources/JoyDOMSampleSpecs/Resources"
let manifestPath = "\(resources)/manifest.json"
let sections = ["textbehavior", "typography"]

func readJSON(_ path: String) -> Any {
    let data = try! Data(contentsOf: URL(fileURLWithPath: path))
    return try! JSONSerialization.jsonObject(with: data, options: [.mutableContainers, .mutableLeaves])
}

func writeJSON(_ obj: Any, to path: String) {
    let data = try! JSONSerialization.data(withJSONObject: obj, options: [.prettyPrinted])
    var s = String(data: data, encoding: .utf8)!
    if !s.hasSuffix("\n") { s += "\n" }
    try! s.write(toFile: path, atomically: true, encoding: .utf8)
}

let manifestObj = readJSON(manifestPath) as! [String: Any]
let samples = manifestObj["samples"] as! [[String: Any]]
var manifestMap: [String: [String: Any]] = [:]
for s in samples {
    if let file = s["file"] as? String {
        manifestMap[file] = (s["snapshot"] as? [String: Any]) ?? [:]
    }
}

struct Stats {
    var total = 0, edited = 0, alreadyBounded = 0, responsive = 0, idAdded = 0, missingManifest = 0
}
var perSection: [String: Stats] = [:]

let fm = FileManager.default

func enumerate(_ dir: String) -> [String] {
    var result: [String] = []
    if let e = fm.enumerator(atPath: dir) {
        while let f = e.nextObject() as? String {
            if f.hasSuffix(".json") {
                result.append("\(dir)/\(f)")
            }
        }
    }
    result.sort()
    return result
}

for section in sections {
    var st = Stats()
    let sectionDir = "\(resources)/\(section)"
    for absPath in enumerate(sectionDir) {
        let rel = String(absPath.dropFirst(resources.count + 1))
        st.total += 1
        guard let snapshot = manifestMap[rel] else {
            st.missingManifest += 1
            FileHandle.standardError.write("  [skip:no-manifest] \(rel)\n".data(using: .utf8)!)
            continue
        }

        let stem = ((rel as NSString).lastPathComponent as NSString).deletingPathExtension
        let sample = readJSON(absPath) as! NSMutableDictionary
        let breakpoints = (sample["breakpoints"] as? [Any]) ?? []
        if stem == "responsive" || !breakpoints.isEmpty {
            st.responsive += 1
            continue
        }

        let layout = (sample["layout"] as? NSMutableDictionary) ?? NSMutableDictionary()
        var propsAny = layout["props"]
        var props = (propsAny as? NSMutableDictionary) ?? NSMutableDictionary()
        var rootId = props["id"] as? String
        var idAdded = false
        if rootId == nil || rootId!.isEmpty {
            rootId = "root"
            let newProps = NSMutableDictionary()
            newProps["id"] = "root"
            for (k, v) in props {
                if let ks = k as? String, ks != "id" {
                    newProps[ks] = v
                }
            }
            props = newProps
            // Rebuild layout dict with key order: type, props, then others, children last
            let oldKeys: [String] = layout.allKeys.compactMap { $0 as? String }
            var orderedKeys: [String] = []
            if oldKeys.contains("type") { orderedKeys.append("type") }
            orderedKeys.append("props")
            for k in oldKeys where k != "type" && k != "props" && k != "children" { orderedKeys.append(k) }
            if oldKeys.contains("children") { orderedKeys.append("children") }
            let snapshotVals: [(String, Any)] = orderedKeys.compactMap { k in
                if k == "props" { return (k, props) }
                if let v = layout[k] { return (k, v) }
                return nil
            }
            layout.removeAllObjects()
            for (k, v) in snapshotVals { layout[k] = v }
            sample["layout"] = layout
            idAdded = true
            st.idAdded += 1
            propsAny = props
        }

        let selector = "#\(rootId!)"
        var style = (sample["style"] as? NSMutableDictionary)
        if style == nil {
            style = NSMutableDictionary()
            // Insert style before breakpoints/layout — rebuild sample
            let oldKeys: [String] = sample.allKeys.compactMap { $0 as? String }
            var orderedKeys: [String] = []
            if oldKeys.contains("version") { orderedKeys.append("version") }
            orderedKeys.append("style")
            for k in oldKeys where k != "version" && k != "style" { orderedKeys.append(k) }
            let snapshotVals: [(String, Any)] = orderedKeys.compactMap { k in
                if k == "style" { return (k, style!) }
                if let v = sample[k] { return (k, v) }
                return nil
            }
            sample.removeAllObjects()
            for (k, v) in snapshotVals { sample[k] = v }
        }
        var rule = (style![selector] as? NSMutableDictionary)
        if rule == nil {
            rule = NSMutableDictionary()
            style![selector] = rule!
        }
        let hasW = rule!["width"] != nil
        let hasH = rule!["height"] != nil
        if hasW && hasH && !idAdded {
            st.alreadyBounded += 1
            continue
        }

        let vw = (snapshot["viewportWidth"] as? Int) ?? 360
        let vh = (snapshot["height"] as? Int) ?? 80

        let newRule = NSMutableDictionary()
        if !hasW {
            let w = NSMutableDictionary()
            w["value"] = vw
            w["unit"] = "px"
            newRule["width"] = w
        } else {
            newRule["width"] = rule!["width"]
        }
        if !hasH {
            let h = NSMutableDictionary()
            h["value"] = vh
            h["unit"] = "px"
            newRule["height"] = h
        } else {
            newRule["height"] = rule!["height"]
        }
        for (k, v) in rule! {
            if let ks = k as? String, ks != "width" && ks != "height" {
                newRule[ks] = v
            }
        }
        style![selector] = newRule

        writeJSON(sample, to: absPath)
        st.edited += 1
    }
    perSection[section] = st
}

print("\n=== Summary ===")
print(String(format: "%-14@ %6@ %7@ %11@ %11@ %9@ %10@",
             "Section" as NSString, "Total" as NSString, "Edited" as NSString,
             "AlrBounded" as NSString, "Responsive" as NSString,
             "IDAdded" as NSString, "NoManifest" as NSString))
for section in sections {
    let st = perSection[section]!
    print(String(format: "%-14@ %6d %7d %11d %11d %9d %10d",
                 section as NSString, st.total, st.edited, st.alreadyBounded, st.responsive, st.idAdded, st.missingManifest))
}

File: scripts/migration/add-root-bounds.py
#!/usr/bin/env python3
"""Merge manifest snapshot width+height into layout-root style rule for non-responsive
textbehavior/typography samples."""
import json
import os
import sys
from collections import OrderedDict

REPO = "/Users/vishnudutt/layout-test"
RESOURCES = os.path.join(REPO, "Sources/JoyDOMSampleSpecs/Resources")
MANIFEST_PATH = os.path.join(RESOURCES, "manifest.json")
SECTIONS = ["textbehavior", "typography"]


def load_manifest():
    with open(MANIFEST_PATH) as f:
        data = json.load(f)
    m = {}
    for entry in data.get("samples", []):
        m[entry["file"]] = entry.get("snapshot", {}) or {}
    return m


def process_sample(rel_path, abs_path, snapshot, stats):
    stem = os.path.splitext(os.path.basename(abs_path))[0]
    with open(abs_path) as f:
        sample = json.load(f, object_pairs_hook=OrderedDict)

    breakpoints = sample.get("breakpoints") or []
    if stem == "responsive" or len(breakpoints) > 0:
        stats["responsive"] += 1
        return False

    layout = sample.get("layout") or {}
    props = layout.get("props") or {}
    root_id = props.get("id")
    id_added = False
    if not root_id:
        root_id = "root"
        if "props" not in layout:
            new_layout = OrderedDict()
            inserted = False
            for k, v in layout.items():
                if k == "children" and not inserted:
                    new_layout["props"] = OrderedDict([("id", "root")])
                    inserted = True
                new_layout[k] = v
            if not inserted:
                new_layout["props"] = OrderedDict([("id", "root")])
            sample["layout"] = new_layout
            layout = new_layout
        else:
            new_props = OrderedDict([("id", "root")])
            for k, v in props.items():
                if k != "id":
                    new_props[k] = v
            layout["props"] = new_props
        id_added = True
        stats["id_added"] += 1

    selector = f"#{root_id}"
    style = sample.get("style")
    if style is None:
        style = OrderedDict()
        sample["style"] = style
    rule = style.get(selector)
    if rule is None:
        rule = OrderedDict()
        style[selector] = rule

    has_w = "width" in rule
    has_h = "height" in rule
    if has_w and has_h and not id_added:
        stats["already_bounded"] += 1
        return False

    vw = snapshot.get("viewportWidth", 360)
    vh = snapshot.get("height", 80)

    if not has_w:
        rule["width"] = OrderedDict([("value", vw), ("unit", "px")])
    if not has_h:
        rule["height"] = OrderedDict([("value", vh), ("unit", "px")])

    ordered = OrderedDict()
    for key in ("width", "height"):
        if key in rule:
            ordered[key] = rule[key]
    for k, v in rule.items():
        if k not in ordered:
            ordered[k] = v
    style[selector] = ordered

    with open(abs_path, "w") as f:
        json.dump(sample, f, indent=2)
        f.write("\n")

    stats["edited"] += 1
    return True


def main():
    manifest = load_manifest()
    overall = {}
    for section in SECTIONS:
        section_dir = os.path.join(RESOURCES, section)
        stats = {"edited": 0, "already_bounded": 0, "responsive": 0,
                 "id_added": 0, "missing_manifest": 0, "total": 0}
        for root, _, files in os.walk(section_dir):
            for name in sorted(files):
                if not name.endswith(".json"):
                    continue
                abs_path = os.path.join(root, name)
                rel = os.path.relpath(abs_path, RESOURCES)
                stats["total"] += 1
                if rel not in manifest:
                    stats["missing_manifest"] += 1
                    print(f"  [skip:no-manifest] {rel}", file=sys.stderr)
                    continue
                snapshot = manifest[rel]
                process_sample(rel, abs_path, snapshot, stats)
        overall[section] = stats

    print("\n=== Summary ===")
    fmt = "{:<14} {:>6} {:>7} {:>11} {:>11} {:>9} {:>10}"
    print(fmt.format("Section", "Total", "Edited", "AlrBounded", "Responsive", "IDAdded", "NoManifest"))
    for s, st in overall.items():
        print(fmt.format(s, st["total"], st["edited"], st["already_bounded"],
                         st["responsive"], st["id_added"], st["missing_manifest"]))


if __name__ == "__main__":
    main()
